### PR TITLE
BUG: groupby sum turning `inf+inf` and `(-inf)+(-inf)` into `nan`

### LINKED ISSUE #53606: BUG: inconsistent treatment of `numpy.Inf` between `groupby.sum()` and `groupby.apply(lambda: _grp: _grp.sum())`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
>>> import numpy as np
>>> import pandas as pd
>>> s = pd.Series([np.Inf, np.Inf, np.Inf])
>>> s.groupby([0, 1, 1]).sum()  # np.Inf treated as NaN
0   inf
1   NaN
dtype: float64

>>> s.groupby([0, 1, 1]).apply(lambda _grp: _grp.sum()) # np.Inf + np.Inf = np.Inf
0   inf
1   inf
dtype: float64
```

### Issue Description

When summing together multiple `np.Inf` values, `groupby.sum()` appears to treat `np.Inf` as `NaN` while `groupby.apply(lambda _grp: _grp.sum())` does not.

### Call Tracing

I traced the calls used in `groupby.sum()`, and verified that the flag [`maybe_use_numba`](https://github.com/pandas-dev/pandas/blob/main/pandas/core/groupby/groupby.py#LL2783C36-L2784C1) evaluated to `False` when calling the [`GroupBy.sum`](https://github.com/pandas-dev/pandas/blob/main/pandas/core/groupby/groupby.py#LL2776) method.  Instead, we delegate to the pandas cython function [`group_sum`](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/groupby.pyx#L666).

As for `groupby,apply()`, we are just [iterating](https://github.com/pandas-dev/pandas/blob/main/pandas/core/groupby/ops.py#LL881) over each of the groups and calling a reduce operation on the pandas Series.  This delegates to the equivalent of `np.nansum`, which treats infinity in a different manner.

### Expected Behavior

Using either `groupby.sum()` or `groupby.apply()` should produce the same result.  I believe that the expected output should be equal to that obtained with `groupby.apply()`, which delegates to calling `np.nansum`:

```
>>> import numpy as np
>>> import pandas as pd
>>> s = pd.Series([np.Inf, np.Inf, np.Inf])
>>> s.groupby([0, 1, 1]).sum()
0   inf
1   inf
dtype: float64
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 965ceca9fd796940050d6fc817707bba1c4f9bff
python           : 3.10.11.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.19.0-42-generic
Version          : #43~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Apr 21 16:51:08 UTC 2
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.2
numpy            : 1.24.3
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 65.5.0
pip              : 23.1.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.13.2
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.10.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None
</details>

### PR DESCRIPTION
**Pull request by Charlie-XIAO**
- [x] Closes #53606
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit)
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file

This modifies `group_sum` in almost the way as #52964 for `group_mean`.

### CODE REVIEW

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -454,8 +454,9 @@ Groupby/resample/rolling
...
 - Bug in :meth:`GroupBy.quantile` may implicitly sort the result index with ``sort=False`` (:issue:`53009`)
 - Bug in :meth:`GroupBy.var` failing to raise ``TypeError`` when called with datetime64, timedelta64 or :class:`PeriodDtype` values (:issue:`52128`, :issue:`53045`)
-- Bug in :meth:`SeriresGroupBy.nth` and :meth:`DataFrameGroupBy.nth` after performing column selection when using ``dropna="any"`` or ``dropna="all"`` would not subset columns (:issue:`53518`)
-- Bug in :meth:`SeriresGroupBy.nth` and :meth:`DataFrameGroupBy.nth` raised after performing column selection when using ``dropna="any"`` or ``dropna="all"`` resulted in rows being dropped (:issue:`53518`)
+- Bug in :meth:`SeriesGroupBy.nth` and :meth:`DataFrameGroupBy.nth` after performing column selection when using ``dropna="any"`` or ``dropna="all"`` would not subset columns (:issue:`53518`)
+- Bug in :meth:`SeriesGroupBy.nth` and :meth:`DataFrameGroupBy.nth` raised after performing column selection when using ``dropna="any"`` or ``dropna="all"`` resulted in rows being dropped (:issue:`53518`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.1.0.rst:458]: Is there an actual difference here? Or is GitHub's diff acting up?

File: pandas/tests/groupby/test_libgroupby.py [hunk 2]
```diff
@@ -302,3 +303,20 @@ def test_cython_group_mean_Inf_at_begining_and_end():
         actual,
         expected,
     )
+
+
+def test_cython_group_sum_Inf_at_begining_and_end():
+    # GH #53606
+    actual = np.array([[np.nan], [np.nan]], dtype="float64")
+    counts = np.array([0, 0], dtype="int64")
+    data = np.array([[np.inf], [np.inf], [np.inf]], dtype="float64")
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_libgroupby.py:312]: Can you also test 

 - (-np.inf) + (-np.inf)
 - np.inf + (-np.inf)
 - (-np.inf) + np.inf

I'd recommend parametrizing the data & expected values

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looks good! A few minor requests.

**Comment by rhshadrach** [comment]
@Charlie-XIAO - small note, we like to use `BUG:` instead of `FIX`. See

https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#finally-make-the-pull-request

for more details.

File: doc/source/whatsnew/v2.1.0.rst (hunk 1 above)
Comment (Charlie-XIAO) [reply doc/source/whatsnew/v2.1.0.rst:458]: Oh I think Series is misspelled here so I changed it.

**Comment by Charlie-XIAO** [comment]
> @Charlie-XIAO - small note, we like to use `BUG:` instead of `FIX`. See
> 
> 
> 
> https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#finally-make-the-pull-request
> 
> 
> 
> for more details.

Oh sure, sorry about that. I mixed up with the prefix of another project lol.

File: doc/source/whatsnew/v2.1.0.rst (hunk 1 above)
Comment (rhshadrach) [reply doc/source/whatsnew/v2.1.0.rst:458]: Ahh, thanks!

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @Charlie-XIAO!

**Pull request merged**